packages/server/graphql/mutations/helpers/notifications/SlackNotificationHelper.ts

```diff
@@ -0,0 +1,178 @@
+import getSummaryText from './getSummaryText'
+import {makeButtons, makeSection, makeSections} from './makeSlackBlocks'
+import {NotificationIntegrationHelper, NotifyResponse} from './NotificationIntegrationHelper'
+import {SlackNotificationEvent} from '../../../../database/types/SlackNotification'
+
+const notifySlack = async (
+  notificationChannel: SlackNotificationAuth,
+  event: SlackNotificationEvent,
+  teamId: string,
+  slackMessage: string | Array<{type: string}>,
+  notificationText?: string
+): Promise<NotifyResponse> => {
+  const {channelId, auth} = notificationChannel
+  const {botAccessToken, userId} = auth
+  const manager = new SlackServerManager(botAccessToken!)
+  const res = await manager.postMessage(channelId!, slackMessage, notificationText)
+  segmentIo.track({
+    userId,
+    event: 'Slack notification sent',
+    properties: {
+      teamId,
+      notificationEvent: event
+    }
+  })
+  if ('error' in res) {
+    const {error} = res
+    if (error === 'channel_not_found') {
+      const r = await getRethink()
+      await r
+        .table('SlackNotification')
+        .getAll(teamId, {index: 'teamId'})
+        .filter({channelId})
+        .update({
+          channelId: null
+        })
+        .run()
+      return {
+        error: new Error('channel_not_found')
+      }
+    } else if (error === 'not_in_channel' || error === 'invalid_auth') {
+      sendToSentry(
+        new Error(`Slack Channel Notification Error: ${teamId}, ${channelId}, ${auth.id}`)
+      )
+      return {
+        error: new Error(error)
+      }
+    }
+  }
+  return 'success'
+}
+
+const makeEndMeetingButtons = (meeting: Meeting) => {
+  const {id: meetingId} = meeting
+  const searchParams = {
+    utm_source: 'slack summary',
+    utm_medium: 'product',
+    utm_campaign: 'after-meeting'
+  }
+  const options = {searchParams}
+  const summaryUrl = makeAppURL(appOrigin, `new-summary/${meetingId}`, options)
+  const makeDiscussionButton = (meetingUrl: string) => ({
+    text: 'See discussion',
+    url: meetingUrl
+  })
+  const summaryButton = {
+    text: 'Review summary',
+    url: summaryUrl
+  } as const
+  switch (meeting.meetingType) {
+    case 'retrospective':
+      const retroUrl = makeAppURL(appOrigin, `meet/${meetingId}/discuss/1`)
+      return makeButtons([makeDiscussionButton(retroUrl), summaryButton])
+    case 'action':
+      const checkInUrl = makeAppURL(appOrigin, `meet/${meetingId}/checkin/1`)
+      return makeButtons([makeDiscussionButton(checkInUrl), summaryButton])
+    case 'poker':
+      const pokerUrl = makeAppURL(appOrigin, `meet/${meetingId}/estimate/1`)
+      const estimateButton = {
+        text: 'See estimates',
+        url: pokerUrl
+      }
+      return makeButtons([estimateButton, summaryButton])
+    default:
+      throw new Error('Invalid meeting type')
+  }
+}
+
+export const SlackNotificationHelper: NotificationIntegrationHelper<SlackNotificationAuth> = (
```

Comment:
I'm still not sure if this is the right thing to do. It is about whether or not we put the responsibility for storing the data into the integration. It should be either the integration knows how to store and retrieve the data or some other part of the app. At the moment it is a mix of both. We put the responsibility for storing clearly into the mutations with the new `IntegrationProvider` and `TeamMemberIntegrationAuth` but the responsibility for retrieving the data is then put into the integration somehow. So I think we have 2 models which we could follow going forward:

1) try to take the structure and storage out of the integrations and manage it uniformly for different ones
  pros:
  * we do that to some extend already with unified auth
  * less duplication in storage/loading code
  cons:
  * storage might not be optimised for single integration, i.e. JSON or lots of NULL fields
 
2) integrations manage the storage on their own
  pros:
  * rest of the app doesn't need to care
  * specifically tailored to the integrations data
  cons:
  * duplication in storing and loading code
  * more possibilities for inconsistent behaviour (e.g. Mattermost notifies only 1 channel while slack can notify multiple and has settings per notification type)

Right now we're doing neither of both which spreads the logic into more places than necessary. I was leaning towards 1) with the caveat that I did not change functionality for how Slack/Mattermost data is loaded, but I would like to have that more consistent going forward.

Reply:
@mattkrick could you take a look?

Reply:
Do you have a code example of each? I'm not 100% sure I understand both solutions.

I'm a big fan of single-purpose classes. My intent with the "Manager" class is something that only facilitates communication with an external API. It can read & write, but it only knows about the service. You should be able to call e.g. `manager.addComment()` and not have to worry whether it'll do that via REST (like jira) or GraphQL (like github). 

Since these managers only care about communication with remote services, I'd like to keep postgres out of them. We can pass their return values to postgres on our own or through an ORM or anything else we choose, but ideally that stuff lives outside of the manager.

All that said, putting a dataloader inside a manager is breaking that encapsulation. Maybe it's OK because it's only a read? Maybe instead we pass in a thunk like `endMeeting(meetingId, getEvents)`. where `getEvents = () => dataLoader.get('slackEvents').load({event: 'meetingEnd'})`?

I'm not sure if that helps. If you write down some code examples we can go through the merits of each option

Reply:
Keeping the dataloader and PG calls out of the managers is basically option 1) which I also wanted to achieve with the version where you've left the comment. Because someone needs to still load the settings/authentication, I had encapsulated it in the `NotificationHelper`. This is similar to `TaskIntegrationManagerFactory` but has the downside, that it needs to know how you load each integration settings. This was amplified by the fact that Slack and Mattermost operate quite differently (from the app logic).

We can also partially solve this by loading all Slack settings for all events (like we've done before). Then we'd only have one initialization step instead of different ones for each method call.